Snapdragon: offset aDSP absolute time

julianoes opened the pull request:
This adds a call on startup of the muorb on the aDSP side to use an
offset for hrt_absolute_call(). This means that the hrt_absolute_call()
on the aDSP (QURT) side should now match the one on the appsproc (Linux) side.

The accuracy still needs to be determined.

Currently I see differences of ~ 30-50 us, but not sure if the check makes too much sense.
```
INFO  time_dsp: 888946621 us, time appsproc: 888946654 us, diff: -33 us
```

Please review @bugobliterator, @kevinmehall 
FYI: @LorenzMeier

kevinmehall commented on src/modules/muorb/krait/px4muorb_KraitRpcWrapper.cpp:
```diff
@@ -59,6 +63,73 @@ static const uint32_t _MAX_TOPICS = 64;
+		return -2;
+	}
+
+	// Do this call right after reading to avoid latency here.W
```
Extra 'W'.

Might be worth repeating the measurement a few times and taking the median in case this code is unlucky enough to get scheduled out right here.

liamstask commented on src/modules/muorb/krait/px4muorb_KraitRpcWrapper.cpp:
```diff
@@ -59,6 +63,73 @@ static const uint32_t _MAX_TOPICS = 64;
+		return -3;
+	}
+
+	time_dsp /= 19.2;
```
where does 19.2 come from?

kevinmehall replied:
Magic number from Qualcomm.

On case 02299793 asked "What is the timer frequency for that value to be converted to standard time units? It seems to count at just under 20MHz. Is there an exact value that I should hard-code into the app, or is there an API to find out the timer frequency?" and they replied with example code that had 19.2 hard-coded into it.

liamstask replied:
gotcha - @julianoes would be good to at least comment it as such for future reference.

julianoes replied:
Will do :).

LorenzMeier commented:
Apart from the comment made by Liam about the magic number its looking good.

kevinmehall commented:
This offset also needs to be used for the DriverFramework timer functions used for sensor timestamping:
https://github.com/PX4/DriverFramework/blob/master/framework/src/DriverFramework.cpp#L205

julianoes commented:
@liamstask, @kevinmehall thanks for the feedback.

I've added a comment to 19.2, and I've changed the timestamping, so `hrt_absolute_time()` is used before publishing the messages instead of the DriverFramework time. I think that's the easiest solution because we don't have to mess with the time functions of the DriverFramework.

LorenzMeier commented:
@julianoes as hotfix that makes perfect sense, but we still need to fix the DriverFramework time base. Could you please file an issue in DriverFramework for that and merge this to master in the meantime?

kevinmehall commented on src/modules/muorb/krait/px4muorb_KraitRpcWrapper.cpp:
```diff
@@ -111,6 +111,9 @@ int calc_timer_diff_to_dsp_us(int32_t *time_diff_us)
 	}
 
+	// The clock count needs to get converted to us. The clock supposedly
+	// runs at just under 20 MHz. The magic value of 19.2 was provided by
```
The "just under 20MHz" probably doesn't need to be mentioned -- that was what I experimentally determined before Qualcomm gave me example code that assumes it's exactly 19.2MHz.

ndepal commented:
Am I correct in assuming that this should provide the same timestamps on the DSP as well as on the linux side, i.e. outside the PX4 main app? Because this is not what I'm observing. When I receive IMU messages via mavlink in a seperate process on the snapdragon, the IMU time stamps start around 3896685466, whereas `gettimeofday()` gives me 1460119337848068, i.e. the time since epoch. Am I just getting the time on the linux side wrong?

ndepal commented on src/modules/muorb/krait/px4muorb_KraitRpcWrapper.cpp:
```diff
@@ -122,13 +195,32 @@ bool px4muorb::KraitRpcWrapper::Initialize()
+
+	int diff = (time_dsp - time_appsproc);
+
+	PX4_INFO("time_dsp: %llu us, time appsproc: %llu us, diff: %d us", time_dsp, time_appsproc, diff);
```
It seems like this test mostly counts the time needed for the get-time calls. Flipping lines 216 and 218 also inverts the sign of `diff`. The average of the two gives a difference of about +9us.

julianoes replied:
Yes that's true but I don't really know how else to verify.

julianoes commented:
@ndepal I think you don't want to use gettimeofday but rather [clock_gettime](http://linux.die.net/man/3/clock_gettime) using `CLOCK_MONOTONIC`.

julianoes commented:
@LorenzMeier I'm scared of changing the clock for the DriverFramework because it would affect the schedule that we don't want to touch once it's running.

I think it's fine to use the timestamps in the wrapper functions, they get called right after the measure call anyway, there is no delay due to context switching, etc.

ndepal commented:
@julianoes that works, thanks!

julianoes commented:
Rebased, and applied.

ndepal commented:
@julianoes have a look at ndepal/Firmware@45b9aa3 . I repeatedly compute the offset in both directions (i.e. getting dsp time first, then appsproc time first). This way I get the reported offset to be around +/-1us (although the biggest factor in this is that I also compute the offset both ways for the check). Let me know if you think this is worth doing and I'll open a new PR.

johnfinley commented on src/drivers/drv_hrt.h:
```diff
@@ -109,6 +109,15 @@ __EXPORT extern hrt_abstime hrt_elapsed_time(const volatile hrt_abstime *then);
+/**
+ * Set a time offset to hrt_absolute_time on the DSP.
+ * @param time_diff_us: time difference of the DSP clock to Linux clock.
+ *   This param is positive because the Linux clock is ahead of the DSP one.
```
It looks like (from other changes) that the parameter does not have to be positive.

julianoes replied:
Oh yes correct! The comment is not up to date. I'll fix it.

julianoes replied:
@johnfinley fixed in https://github.com/PX4/Firmware/pull/4217/commits/fbd7c73043b184fbae0e94f3a2053cafdd89a145
